### DISCUSSION TITLE
Pagination in a method other than render

### DISCUSSION THREAD

**Discussion by fornaserofranco** [livewire/livewire#5175] [1. Help]
I am developing a component in which there is a search for items inside it. I have a form in that component that calls the submit() method which fetches objects and returns them to the view. The problem is that the number of elements can be variable and for this I require a pagination, but I can't make it work. It doesn't give me any errors or anything, it just doesn't show the results.

This is the controller that calls the view that contains the component:
```php
 public function editar(Pedido $pedido)
 {
    return View('modificar', compact('pedido'));
 }
```

This is the view
```blade
@extends('layouts.app')

@section('content')
    @livewire('modificar-pedido', ['pedido' => $pedido])
@endsection
```

This is the part of my component that does the pagination
```blade
<div class="card-body">
            <form wire:submit.prevent="submit(Object.fromEntries(new FormData($event.target)))">
                @csrf
                <div class="d-flex justify-content-between mx-auto  mt-5 fs-14 row">
                    <select class="form-select px-0 rounded btn btn-secondary mb-3 col-lg-3 pr-lg-3 text-center"
                        name="filtro" id="filtro">
                        <option
                            @isset($filtro) selected {{ $filtro == 'codigo_barras' ? 'selected' : '' }} @endisset
                            value="codigo_barras">Codigo de barras</option>
                        <option
                            @isset($filtro) {{ $filtro == 'codigo_interno' ? 'selected' : '' }} @endisset
                            value="codigo_interno">Codigo interno</option>
                        <option
                            @isset($filtro) {{ $filtro == 'detalle' ? 'selected' : '' }} @endisset
                            value="detalle">Detalle</option>
                        <option @isset($filtro) {{ $filtro == 'marca' ? 'selected' : '' }} @endisset
                            value="marca">Marca</option>
                    </select>

                    <input type="text" class="form-control border-0 mb-3 col-lg-5 mx-lg-3" placeholder=""
                        name="input" value="@isset($input) {{ $input }} @endisset">

                    <input class="btn btn-secondary col-lg-3 mb-3" type="submit" id="button_busqueda"
                        value="Buscar"></a>
                </div>
            </form>
            @isset($productos)
                @if (count($productos) == 0)
                    <div class="alert alert-warning w-75 mx-auto">
                        No encontramos registros que coincidan con su búsqueda.
                    </div>
                @else
                    <div class="alert alert-info w-75 mx-auto">
                        Recuerde que los precios incluyen IVA.
                    </div>
                    <table class="table mt-5 mb-5 w-100 mx-auto fs-8 fs-md-14" style="max-width: 1425px"
                        id="table_productos">
                        <thead class="thead-dark">
                            <tr>
                                <th scope="col">Código Interno</th>
                                <th scope="col">Código de Barras</th>
                                <th scope="col">Marca</th>
                                <th scope="col">Detalle</th>
                                <th scope="col">Precio</th>
                            </tr>
                        </thead>
                        <tbody>
                            @foreach ($productos as $key => $producto)
                                <tr>
                                    <td class="py-1 py-md-2">{{ $producto->codigo }}</td>
                                    <td class="py-1 py-md-2">{{ $producto->coddun14 }}</td>
                                    <td class="py-1 py-md-2">{{ $producto->marca_cor }}</td>
                                    <td class="py-1 py-md-2">{{ $producto->detalle }}</td>
                                    <td class="py-1 py-md-2">$ {{ number_format($producto->precio, 2, ',', '.') }}</td>
                                </tr>
                            @endforeach
                        </tbody>
                    </table>
                @endif
                {{$productos->links()}}
            @endisset
</div>
```

This is my component controller
```php
<?php

namespace App\Http\Livewire;

use Livewire\Component;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Validator;
use Livewire\WithPagination;

class ModificarPedido extends Component
{

    use WithPagination;

    public $pedido;
    public $id_producto;
    public $productos;

    public function render()
    {
        return view('livewire.modificar-pedido', ['pedido' => $this->pedido]);
    }

    public function eliminarProducto($id_producto)
    {
        $this->pedido->detallePedido()->where('id_producto', $id_producto)->delete();
        $this->pedido->refresh();
    }

    public function submit($formData)
    {
        
        if ($formData['filtro'] == 'codigo_barras' || $formData['filtro'] == 'codigo_interno') {
            Validator::make($formData, [
                'input' => 'required|numeric',
            ]);
        } else {
            Validator::make($formData, [
                'input' => 'required|string'
            ]);
        }

        $query =  DB::table('vista_base_listados');
        //filters bla bla bla

        $this->productos = $query->get();

        return view('livewire.modificar-pedido', ['productos' => $this->productos]);
    }
}
```
Things to keep in mind:

- I already made a livewire component in this way, doing a DB::table() query and I didn't have any problem. The only thing different was that everything was done in the render() method.
- The code I am showing is the one that is currently working, albeit without pagination.
- I tried to do $query->paginate(10) and return like this: return view('livewire.modify-order', ['products' => $query->paginate(10)]). But it just doesn't show anything.
- The way of bringing me the products is fine, if I do a dd() I see the results and they are correct.
- If I do a dd() of $products in the component, the result I get is null

**Comment by Martleby** [comment]
Hello,

From my understanding, when you return from your submit method you're returning control back to Livewire's lifecycle pipeline, rather than forcing your view to re-render. Livewire then proceeds through its lifecycle and eventually gets to the render method.  Since productos aren't passed from the render method to the view your @isset will always be false.

There are a lot of different ways to achieve what you want though, here's a pattern I like to use:

1) Instead of passing $formData to your submit method, I would bind these to properties on your component using wire:model.

2) I would also recommend using Livewire's validation.

Then my component might look a bit like this:

```php
class ModificarPedido extends Component
{

    use WithPagination;

    public $pedido;
    public $id_producto;
    public $productos;
    public $filtro;
    public $readyToSearch = false;

    protected $rules = [... rules here ...];


    public function render()
    {
        return view('livewire.modificar-pedido', ['pedido' => $this->pedido])
             ->with('productos', $this->readyToSearch ? $this->getProductos()->paginate(10) : []);
    }

    public function eliminarProducto($id_producto)
    {
        $this->pedido->detallePedido()->where('id_producto', $id_producto)->delete();
        $this->pedido->refresh();
    }

    public function submit($formData)
    {
       $this->validate();
      //If we pass validation, set ready to search to true
       $this->readyToSearch = true; 
    }

    //If any of the form filters change, set ready to search to false
    public function updating()
    {
       $this->readyToSearch = false;
    }

    public function getProductos()
    {
         return DB::table('vista_base_listados');
        //filters bla bla bla
     }
}

```

Hope that's of some use to you.

Cheers,
Martin.

**Reply by fornaserofranco** [reply]
@Martleby I spend some time on this. I'm testing it and I'm having a problem.
![image](https://user-images.githubusercontent.com/108434960/185408337-1a959b7f-d05f-4cde-93ce-d99c74b243b5.png)
If I do a dd in the livewire controller if I see the elements that I need correctly, but on the template side this happens. Any ideas?

**Reply by Martleby** [reply]
Hmmm. Seems like the connection object is being passed to the view, rather than the results. Can you share your code?

**Reply by fornaserofranco** [reply]
@Martleby Made a new post here: https://github.com/livewire/livewire/discussions/5189

I've been learning a lot about pagination and filters on livewire these days. Although I didn't know how to solve the original problem, I was continuing with other things. I'll probably pick it up again in the afternoon.

**Reply by fornaserofranco** [reply]
@Martleby 
I was finally able to get pagination to work the way I want. It cost a lot... and the worst thing is that the mistake was, after all, quite simple.

The problem was the declaration of the public variable $products. I tried in a thousand ways to do the pagination and always $productos was passed as null. Whatever I did, I was getting null in my view, even if I defined it as a text string. The problem was that damn declared variable. As much as $products is being passed and not $this->productos in view, $this->productos arrived.

I don't know why livewire had this behavior, maybe it's a library error or I still don't fully understand how it works.

Actually, my code had many modifications since I made the post, but I can say without a doubt that your comment helped me, since I used wire:model for the search form and that made my job easier.
